src/process_quantiles.R

```diff
@@ -0,0 +1,164 @@
+
+########## Functions ########## 
+task_id_to_hru_seq <- function(task_id, n_hrus_per_task = 1000) {
+  # Convert task id to HRU ids (increment by n_hrus_per_task)
+  hru_id_start <- ((task_id-1)*n_hrus_per_task + 1) # which hru to start with
+  hru_id_end <- hru_id_start + (n_hrus_per_task-1) # which hru to end with
+  
+  # If we are past the last column, cut the final col id off
+  n_hrus <- 109951
+  hru_id_end <- ifelse(hru_id_end > n_hrus,
+                       yes = n_hrus,
+                       no = hru_id_end)
+  
+  hru_seq <- hru_id_start:hru_id_end
+  return(hru_seq)
+}
+read_ncdf_data <- function(fn, varid, hru_seq) {
+  
+  # The file is NetCDF
+  nc <- nc_open(fn)
+  
+  message(sprintf("Reading in NetCDF data for %s", varid))
+  # Only load rows for current HRUs
+  data_nc <- ncvar_get(nc, varid, 
+                       start = c(head(hru_seq, 1), 1),
+                       count = c(length(hru_seq), -1))
+  time <- ncvar_get(nc, "time")
+  hruid <- ncvar_get(nc, "hruid")[hru_seq]
+  
+  # Convert ncdf4 times to R dates
+  time_att <- ncdf4::ncatt_get(nc, "time")
+  time_start <- as.Date(gsub("days since ", "", time_att$units))
+  time_fixed <- time + time_start # creates dates from the "days since" var
+  
+  # Keep only complete years of data
+  # data_years <- format(time_fixed, "%Y")
+  # n_days_per_year <- table(data_years)
+  # data_years_complete <- names(n_days_per_year)[n_days_per_year >= 365]
+  
+  year_doy_vector <- format(time_fixed, "%Y_%j")
+  dt <- as.data.table(data_nc)
+  dt[, hruid := hru_seq]
+  names(dt) <- c(year_doy_vector, "hruid")
+  dt_long <- melt(dt, id.vars = "hruid", variable.name = "year_doy")
+  
+  return(dt_long)
+}
+########## ////// ########## 
+
+# Identify task id from yeti environment & convert to HRU ids ----
+# task_id <- as.numeric(Sys.getenv('SLURM_ARRAY_TASK_ID', 'NA'))
+task_id <- 2 # Will uncomment once setup with slurm
+hru_seq <- task_id_to_hru_seq(task_id)
+hru_start <- head(hru_seq, 1)
+hru_end <- tail(hru_seq, 1)
+quantile_fn <- sprintf("grouped_quantiles/total_storage_quantiles_%s_to_%s.csv", hru_start, hru_end)
+
+if(!file.exists(quantile_fn)) {
+
+  message(sprintf("Started this task at %s", Sys.time()))
+
+  vars <- c(
+    "soil_moist_tot",
+    "pkwater_equiv",
+    "hru_intcpstor", 
+    "hru_impervstor",
+    "gwres_stor", 
+    "dprst_stor_hru"
+  )
+  
+  dt_long <- c()
+  for(var in vars) {
+    # Read in the two datasets
+    ncdf_fn <- sprintf("historical_%s_out.nc", var)
+    var_df <- read_ncdf_data(ncdf_fn, var, hru_seq)
+    dt_long <- rbind(dt_long, var_df)
+  }
+  
+  # Split year_doy into separate columns
+  # Converting to numeric (type_convert = TRUE) slows down
+  dt_long[, c("year", "doy") := tstrsplit(year_doy, "_", fixed=TRUE)]
+  
+  # `fill = NA` to account for missing dates at beginning 
+  #   of 1980 and end of 2019
+  # This spreads + sums the values! Takes ~ 10 sec
+  dt_wide <- dcast(dt_long, hruid+year ~ doy, fun=sum, fill=NA)
```

Comment:
Ohhh, this is where you're summing the 6 values! When I reviewed this yesterday I wondered why you were using fun=sum rather than, say, fun=c, but now I finally understand. I still wonder if `dt_long = dt_long + var_df` in the above variables loop could be even faster (and less memory intensive) than waiting until this line to sum the variables, but it does seem that dcast is pretty darn fast.

Reply:
ahhhhh yes. OK, I'll work on some changes for this after the class this evening. I've already got a bit going that I have worked on during lunch but won't have a chance to finish. Should speed it up even more but I feel pretty good about the 4 min time for each group of 1000 HRUs now so @mhines-usgs doesn't need to wait necessarily.

Reply:
I'll right! I've got a fix for this. Brings the time down to 3.88 min :)